---
title: "MED264: Rentrez Tutorial"
author: "Tim Dennis"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: yes
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = FALSE,
  comment = "#"
)
knitr::opts_knit$set(upload.fun = knitr::imgur_upload, base.url = NULL)  

```

```{r, count_recs, echo=FALSE}
library(rentrez)
count_recs <- function(db, denom) {
    nrecs <-  rentrez::entrez_db_summary(db)["Count"]
    round(as.integer(nrecs)/denom, 1)
}
```
## Take the QUIZ! 

Go to: 

1. <https://b.socrative.com/login/student/>
2. USE the room number: QTKGZXF6G

## NCBI

NCBI has a lot of data in it.  As of today, it has: 

* `r count_recs("pubmed",1e6)` million papers in [PubMed](http://www.ncbi.nlm.nih.gov/pubmed/),
* includes `r count_recs("pmc", 1e6)` million full-text records available in [PubMed Central](http://www.ncbi.nlm.nih.gov/pubmed/)
*  [The NCBI Nucleotide Database](http://www.ncbi.nlm.nih.gov/nuccore) (which includes GenBank) has data for `r count_recs("nuccore", 1e6)` million different sequences
* [dbSNP](http://www.ncbi.nlm.nih.gov/snp/) describes `r count_recs("snp", 1e6)` million different genetic variants

All records can be cross-referenced with the  `r round(entrez_search(db="taxonomy", term='species[RANK]')$count/1e6,2)` million species in the [NCBI taxonomy](www.ncbi.nlm.nih.gov/taxonomy) or `r count_recs("omim", 1e3)` thousand disease-associated records in [OMIM](http://www.ncbi.nlm.nih.gov/omim).

##rentrez

`rentrez` provides functions that work with the [NCBI Eutils](http://www.ncbi.nlm.nih.gov/books/NBK25500/)
API to search, download data from, and otherwise interact with NCBI databases.

```
library(devtools)
install_github("ropensci/rentrez")
```
##rentrez Uses the EUtils API

* Read more about [EUtils](http://www.ncbi.nlm.nih.gov/books/NBK25500/)
* Working with the EUtils API will often require making multiple calls using the entrez package.
* Python has a module with similar functionality in the Biopython module; http://biopython.org/wiki/Main_Page

## Gettting started with rentrez 

### Install

From the R Cran repository: 

```{r eval=FALSE}
install.packages('rentrez')
```

From the development branch: 

```{r eval=FALSE}
install_github("ropensci/rentrez")
```

`library()` tells our R environment to load the package for use.

```{r}
library(rentrez)
```

We can use `entrez_dbs` to get a list of the databases we can search. 

```{r}
entrez_dbs()
```


**Functions that help you learn about NCBI databases**

| Function name            | Return                                               |
|--------------------------|------------------------------------------------------|
| `entrez_db_summary()`    | Brief description of what the database is            |
| `entrez_db_searchable()` | Set of search terms that can used with this database |
| `entrez_db_links() `     | Set of databases that might contain linked records   |


Let's retrieve a description of on of the databases: 

```{r}
entrez_db_summary('dbvar')
#entrez_db_summary('snp')
```

We can also use `entrez_db_searchable` to see what search fields and qualifiers are allowable:

```{r}
entrez_db_searchable("sra")
```

* Check out <https://www.ncbi.nlm.nih.gov/entrez/query/static/help/Summary_Matrices.html#Search_Fields_and_Qualifiers> for more information on the search term **qualifiers**. 

##A Simple Example: Find by DOI

**Use a DOI to return the PMID of an article using `entrez_search`** 

Use an article DOI: 
[Cancer risk reduction and reproductive concerns in female BRCA1/2 mutation carriers](http://dx.doi.org/10.1007/s10689-007-9171-7). DOI of 10.1007/s10689-007-9171-7. 

```{r}
library(rentrez)
#wcancer_paper <- entrez_search(db="pubmed", term="10.1007/s10689-007-9171-7[doi]")
wcancer_paper <- entrez_search(db="pubmed", term="10.1002/ijc.21536[doi]")

wcancer_paper$ids
```

**Get some summary info**
```{r}
wcan_summary <- entrez_summary(db="pubmed", wcancer_paper$ids)
wcan_summary$title
wcan_summary$authors
```

**[Challenge](exercises.html)** (5min)

## Search Terms:  `entrez_search()`

* More often what you'll want to do with `rentrez` is search a given NCBI database to find records that match some keywords
* We use the function `entrez_search()` to do this
* Simplest case you just need to provide a database name (`db`) and a search term (`term`) so let's search PubMed for articles about the `Natural Language Processing`:

```{r eg_search}
r_search <- entrez_search(db="pubmed", term="Natural Language Processing")
```

* The object returned by a search acts like a `list`, and you can get a summary of  its contents by printing it.

```{r print_search}
r_search
```

* NCBI's server has worked out that we meant NPL, and so included the ['MeSH' term](http://www.ncbi.nlm.nih.gov/mesh) 
* there are many more 'hits' for this search than there are unique IDs contained in this object. That's because the optional argument `retmax`, which controls the maximum number of returned values has a default
value of 20.
* The IDs are the most important thing returned here. They
allow us to fetch records matching those IDs, gather summary data about them or find
cross-referenced records in other databases. 
* We access the IDs as a vector using the `$` operator:

```{r search_ids}
r_search$ids
```

* let's set the retmax up to retrieve more ids

```{r searchids_2}
another_r_search <- entrez_search(db="pubmed", term="Natural Language Processing", retmax=40)
another_r_search
```

* If we want to get IDs for all of the thousands of records that match this
search, we can use the NCBI's web history feature. 

## Using search field operators

NCBI has search field operators that we can add to queries `query[field]`. 

* For instance, we can find next generation sequence datasets for the (amazing...) ciliate
_Tetrahymena thermophila_ by using the organism ('ORGN') search field:

```{r, Tt}
entrez_search(db="sra",
              term="Tetrahymena thermophila[ORGN]",
              retmax=0)
```

We can narrow our focus to only those records that have been added recently (using the colon to 
specify a range of values):


```{r, Tt2}
entrez_search(db="sra",
              term="Tetrahymena thermophila[ORGN] AND 2013:2015[PDAT]",
              retmax=0)
```
Or include recent records for either _T. thermophila_ or it's close relative _T.
borealis_ (using parentheses to make ANDs and ORs explicit).


```{r, Tt3}
entrez_search(db="sra",
              term="(Tetrahymena thermophila[ORGN] OR Tetrahymena borealis[ORGN]) AND 2013:2015[PDAT]",
              retmax=0)
```

The set of search terms available varies between databases. You can get a list
of available terms or any given data base with `entrez_db_searchable()`

```{r, sra_searchable}
entrez_db_searchable("sra")
```
##Precise queries using MeSH terms

* NCBI allows searches using [Medical Subject Heading (MeSH)](http://www.ncbi.nlm.nih.gov/mesh) terms
   * these terms create a 'controlled vocabulary',  and allow users to make very finely controlled queries of databases.
* CASE: if you were interested in reviewing studies on how a class of anti-malarial drugs called Folic Acid Antagonists work against _Plasmodium vivax_ (a particular species of malarial parasite), you could use this search:

```{r, mesh}
entrez_search(db   = "pubmed",
              term = "(vivax malaria[MeSH]) AND (folic acid antagonists[MeSH])")
```

* MeSH terms are available as a database from the NCBI
* You can download detailed information about each term and findthe ways in which terms relate to each other using `rentrez`. 
* Search specific terms with `entrez_search(db="mesh", term =...)` and learn about the results of your search using the tools described below.

* **[Counting - use Ted's example](2_entrez_search.html)**

## Linking docs: `entrez_link()`

* One of the strengths of the NCBI databases is that records of one type are connected to  other records within the NCBI or to external data sources
*`entrez_link()` allows users to discover these links between records.
* Let's find all NCBI data associated with a single gene (in this case the  Amyloid Beta Precursor gene, the product of which is associated with the  plaques that form in the brains of  Alzheimer's Disease patients).
* we need to provide an ID (`id`), the database from which this ID comes (`dbfrom`) and the name of a database in which to find linked records (`db`)
* If we set this last argument to 'all' we can find links in multiple databases:

```{r elink0}
all_the_links <- entrez_link(dbfrom='gene', id=351, db='all')
all_the_links
```
* return object behaves like a list
```{r elink_link}
all_the_links$links
```
* names of the list elements are in the format `[source_database]_[linked_database]`  and the elements themselves contain a vector of linked-IDs
* if we want to find open access publications associated with this gene we could get linked records
in PubMed Central:

```{r, elink_pmc}
all_the_links$links$gene_pmc[1:10]
```

Or if were interested in this gene's role in diseases we could find links to clinVar:

```{r, elink_omim}
all_the_links$links$gene_clinvar

```

## Where to learn more???

* Most of this tutorial was based on the [ROpenscience tutorial](https://ropensci.org/tutorials/rentrez_tutorial.html) on rentrez. 
* I also used some of Ted Hart's [Rentrez demo](https://github.com/ropensci/workshops-scio14-2014-02) as part of SIO14. 