---
title: "MED264: Rentrez Tutorial"
author: "Reid Otsuji adpted from Tim Dennis MED264"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: yes
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = FALSE,
  comment = "#"
)
knitr::opts_knit$set(upload.fun = knitr::imgur_upload, base.url = NULL)  

```

```{r, count_recs, echo=FALSE}
library(rentrez)
count_recs <- function(db, denom) {
    nrecs <-  rentrez::entrez_db_summary(db)["Count"]
    round(as.integer(nrecs)/denom, 1)
}
```
<!--## Take the QUIZ! 

Go to: 

1. <https://b.socrative.com/login/student/>
2. USE the room number: QTKGZXF6G
-->

## NCBI Intro

NCBI has a lot of data in it.  As of today, it has: 

* `r count_recs("pubmed",1e6)` million papers in [PubMed](http://www.ncbi.nlm.nih.gov/pubmed/),
* includes `r count_recs("pmc", 1e6)` million full-text records available in [PubMed Central](http://www.ncbi.nlm.nih.gov/pubmed/)
*  [The NCBI Nucleotide Database](http://www.ncbi.nlm.nih.gov/nuccore) (which includes GenBank) has data for `r count_recs("nuccore", 1e6)` million different sequences
* [dbSNP](http://www.ncbi.nlm.nih.gov/snp/) describes `r count_recs("snp", 1e6)` million different genetic variants

All records can be cross-referenced with the  `r round(entrez_search(db="taxonomy", term='species[RANK]')$count/1e6,2)` million species in the [NCBI taxonomy](www.ncbi.nlm.nih.gov/taxonomy) or `r count_recs("omim", 1e3)` thousand disease-associated records in [OMIM](http://www.ncbi.nlm.nih.gov/omim).

## rentrez package

`rentrez` provides functions that work with the [NCBI Eutils](http://www.ncbi.nlm.nih.gov/books/NBK25500/)
API to search, download data from, and otherwise interact with NCBI databases.

```{r eval=FALSE}
library(devtools)
install_github("ropensci/rentrez")
```


##rentrez Uses the EUtils API

* Read more about [EUtils](http://www.ncbi.nlm.nih.gov/books/NBK25500/)
* Working with the EUtils API will often require making multiple calls using the entrez package.
* Python has a module with similar functionality in the Biopython module; http://biopython.org/wiki/Main_Page

## Gettting started with rentrez 

### Install

* Install from the R Cran repository: 

```{r eval=FALSE}
install.packages('rentrez')
```

* From the development branch on GitHub: 

```{r eval=FALSE}
install_github("ropensci/rentrez")
```

* `library()` tells our R environment to load the package for use.

```{r}
library(rentrez)
```


## Getting started
* Let's take a look at the available NCBI databases
* We can use `entrez_dbs` to get a list of the databases we can search. 

```{r}
entrez_dbs()
```


**Helper Functions that help you learn about NCBI databases**

| Function name            | Return                                               |
|--------------------------|------------------------------------------------------|
| `entrez_db_summary()`    | Brief description of what the database is            |
| `entrez_db_searchable()` | Set of search terms that can used with this database |
| `entrez_db_links() `     | Set of databases that might contain linked records   |


* Let's retrieve a description of on of the databases: 

```{r}
entrez_db_summary('dbvar')
#entrez_db_summary('snp')
```

* We can also use `entrez_db_searchable` to see what search fields and qualifiers are allowable
* For example, we can see what searchable fields (qualifiers) can be used with the Sequence Read Archive (SRA) database:

```{r}
entrez_db_searchable("sra")
```

* Check out <https://www.ncbi.nlm.nih.gov/entrez/query/static/help/Summary_Matrices.html#Search_Fields_and_Qualifiers> for more information on the search term **qualifiers**. 




## Search Terms:  `entrez_search()`

* More often what you'll want to do with `rentrez` is search a given NCBI database to find records that match some keywords
* We use the function `entrez_search()` to do this
* Simplest case you just need to provide a database name (`db`) and a search term (`term`) so let's search PubMed for articles about the `Natural Language Processing`:

```{r eg_search}
r_search <- entrez_search(db="pubmed", term="Natural Language Processing")
```

* The object returned by a search acts like a `list`, and you can get a summary of  its contents by printing it.

```{r print_search}
r_search
```

* NCBI's server has worked out that we meant NPL, and so included the ['MeSH' term](http://www.ncbi.nlm.nih.gov/mesh) 
* there are many more 'hits' for this search than there are unique IDs contained in this object. That's because the optional argument `retmax`, which controls the maximum number of returned values has a default
value of 20.
* The IDs are the most important thing returned here. They
allow us to fetch records matching those IDs, gather summary data about them or find
cross-referenced records in other databases. 
* We access the IDs as a vector using the `$` operator:

```{r search_ids}
r_search$ids
```

* let's set the retmax up to retrieve more ids

```{r searchids_2}
another_r_search <- entrez_search(db="pubmed", term="Natural Language Processing", retmax=40)
another_r_search
```

* If we want to get IDs for all of the thousands of records that match this
search, we can use the NCBI's web history feature. 


## A Simple Example: Find by DOI using `entrez_search()`

**Use a DOI to return the PMID of an article using `entrez_search`** 

Use an article DOI: 
[Cancer risk reduction and reproductive concerns in female BRCA1/2 mutation carriers](http://dx.doi.org/10.1007/s10689-007-9171-7). DOI of 10.1007/s10689-007-9171-7. 

```{r}
wcancer_paper <- entrez_search(db="pubmed", term="10.1002/ijc.21536[doi]")

wcancer_paper$ids
```

**Get some summary info**
```{r}
wcan_summary <- entrez_summary(db="pubmed", wcancer_paper$ids)
wcan_summary$title
wcan_summary$authors
```


## Using search field operators

* NCBI has search field operators that we can add to queries `query[search field]`. 

* For instance, we can find next generation sequence datasets for the (amazing...) ciliate
_Tetrahymena thermophila_ by using the organism ('ORGN') search field:

```{r, Tt}
entrez_search(db="sra",
              term="Tetrahymena thermophila[ORGN]",
              retmax=0)
```

* We can narrow our focus to only those records that have been added recently (using the colon to 
specify a range of values):


```{r, Tt2}
entrez_search(db="sra",
              term="Tetrahymena thermophila[ORGN] AND 2013:2015[PDAT]",
              retmax=0)
```
* Or include recent records for either _T. thermophila_ or it's close relative _T.
borealis_ (using parentheses to make ANDs and ORs explicit).


```{r, Tt3}
entrez_search(db="sra",
              term="(Tetrahymena thermophila[ORGN] OR Tetrahymena borealis[ORGN]) AND 2013:2015[PDAT]",
              retmax=0)
```

* The set of search terms available varies between databases. You can get a list
of available terms or any given data base with `entrez_db_searchable()`

```{r, sra_searchable}
entrez_db_searchable("sra")
```
## Precise queries using MeSH terms

* NCBI allows searches using [Medical Subject Heading (MeSH)](http://www.ncbi.nlm.nih.gov/mesh) terms
   * these terms create a 'controlled vocabulary',  and allow users to make very finely controlled queries of databases.
* CASE: if you were interested in reviewing studies on how a class of anti-malarial drugs called Folic Acid Antagonists work against _Plasmodium vivax_ (a particular species of malarial parasite), you could use this search:

```{r, mesh}
entrez_search(db   = "pubmed",
              term = "(vivax malaria[MeSH]) AND (folic acid antagonists[MeSH])")
```

* MeSH terms are available as a database from the NCBI
* You can download detailed information about each term and findthe ways in which terms relate to each other using `rentrez`. 
* Search specific terms with `entrez_search(db="mesh", term =...)` and learn about the results of your search using the tools described below.


## **[Challenge](exercises.html)** (5min) 


## * **[Section 2 - Search and plot trendy topics in genetics example](02_entrez_search.html)**

## Linking docs: `entrez_link()`

* One of the strengths of the NCBI databases is that records of one type are connected to  other records within the NCBI or to external data sources

*`entrez_link()` allows users to discover these links between records.

* Let's find all NCBI data associated with a single gene (in this case the  Amyloid Beta Precursor gene, the product of which is associated with the  plaques that form in the brains of  Alzheimer's Disease patients).

* we need to provide an ID (`id`), the database from which this ID comes (`dbfrom`) and the name of a database in which to find linked records (`db`)

* If we set this last argument to 'all' we can find links in multiple databases:

```{r elink0}
all_the_links <- entrez_link(dbfrom='gene', id=351, db='all')
all_the_links
```
* return object behaves like a list

```{r elink_link}
all_the_links$links
```
* names of the list elements are in the format `[source_database]_[linked_database]`  and the elements themselves contain a vector of linked-IDs

* if we want to find open access publications associated with this gene we could get linked records
in PubMed Central:

```{r, elink_pmc}
all_the_links$links$gene_pmc[1:10]
```

* Or if were interested in this gene's role in diseases we could find links to clinVar:

```{r, elink_omim}
all_the_links$links$gene_clinvar

```

# original lesson ended here - go to lesson 2 to show plotting if less time


## Narrowing our focus

* If we know beforehand what sort of links we’d like to find, we can to use the `db` argument to narrow the focus of a call to `entrez_link()`.

* For instance, say we are interested in knowing about all of the RNA transcripts associated with the Amyloid Beta Precursor gene in humans. 

* Transcript sequences are stored in the nucleotide database (referred to as nuccore in EUtils), so to find transcripts associated with a given gene we need to set dbfrom=gene and db=nuccore.



```{r}
nuc_links <- entrez_link(dbfrom='gene', id=351, db='nuccore')
nuc_links
```

```{r}
nuc_links$links
```

* The object we get back contains links to the nucleotide database generally, but also to special subsets of that database like `refseq`. 

* We can take advantage of this narrower set of links to find IDs that match unique transcripts from our gene of interest.


```{r}
nuc_links$links$gene_nuccore_refseqrna
```

* We can use these ids in calls to `entrez_fetch()` or `entrez_summary()` to learn more about the transcripts they represent.


## External links
* In addition to finding data within the NCBI, `entrez_link` can turn up connections to external databases. Perhaps the most interesting example is finding links to the full text of papers in PubMed. 

* For example, when I wrote this document the first paper linked to `Amyloid Beta Precursor` had a unique ID of `25500142`. We can find links to the full text of that paper with `entrez_link` by setting the cmd argument to `‘llinks’`:

```{r}
paper_links <- entrez_link(dbfrom="pubmed", id=25500142, cmd="llinks")
paper_links
```

* Each element of the `linkouts` object contains information about an external source of data on this paper:


```{r}
paper_links$linkouts
```
* Each of those linkout objects contains quite a lot of information, but the URL is probably the most useful. 

* For that reason, `rentrez` provides the function `linkout_urls` to make extracting just the URL simple:

```{r}
linkout_urls(paper_links)
```

* The full list of options for the cmd argument are given in in-line documentation `(?entrez_link)`. 

* If you are interested in finding full text records for a large number of articles checkout the package [fulltext](https://github.com/ropensci/fulltext) which makes use of multiple sources (including the NCBI) to discover the `full text` articles.

## Using more than one ID 
* It is possible to pass more than one ID to `entrez_link()`. 

* By default, doing so will give you a single elink object containing the complete set of links for all of the IDs that you specified. 

* So, if you were looking for protein IDs related to specific genes you could do:
 
```{r}
all_links_together <- entrez_link(db="protein", dbfrom="gene", id=c("93100", "223646"))
all_links_together
```

```{r}
all_links_together$links$gene_protein
```

* Although this behaviour might sometimes be useful, it means we’ve lost track of which `protein` ID is linked to which `gene` ID. 

* To retain that information we can set `by_id` to TRUE. This gives us a list of elink objects, each once containing links from a single gene ID:

```{r}
all_links_sep <- entrez_link(db="protein", dbfrom="gene", id=c("93100", "223646"), by_id = TRUE)
all_links_sep
```

```{r}
lapply(all_links_sep, function(x) x$links$gene_protein)
```

## Getting summary data: entrez_summary()
* Having found the unique IDs for some records via `entrez_search` or `entrez_link()`, you are probably going to want to learn something about them. 

* The Eutils API has two ways to get information about a record. 

* `entrez_fetch()` returns ‘full’ records in varying formats and `entrez_summary()` returns less information about each record, but in relatively simple format. 

* Very often the summary records have the information you are after, so rentrez provides functions to parse and summarise summary records.

## The summary record
* `entrez_summary()` takes a vector of unique IDs for the samples you want to get summary information from.

* Let’s start by finding out something about the paper describing `[Taxize](https://github.com/ropensci/taxize)`, using its PubMed ID:

```{r}
taxize_summ <- entrez_summary(db="pubmed", id=24555091)
taxize_summ
```
* Once again, the object returned by `entrez_summary` behaves like a list, so you can extract elements using $. 

* For instance, we could convert our PubMed ID to another article identifier…

```{r}
taxize_summ$articleids
```

* …or see how many times the article has been cited in PubMed Central papers

```{r}
taxize_summ$pmcrefcount
```

## Dealing with many records

* If you give `entrez_summary()` a vector with more than one ID you’ll get a list of summary records back.

* Let’s get those `Plasmodium vivax` papers we found in the `entrez_search()` section back, and fetch some summary data on each paper:

```{r}
vivax_search <- entrez_search(db = "pubmed", 
                              term = "(vivax malaria[MeSH]) AND (folic acid antagonists[MeSH])")
multi_summs <- entrez_summary(db="pubmed", id=vivax_search$ids)
```

* `rentrez` provides a helper function, `extract_from_esummary()` that takes one or more elements from every summary record in one of these lists. 

* Here it is working with one…

```{r}
extract_from_esummary(multi_summs, "fulljournalname")
```

* and several elements
(using knitr package used for dynamic report generation to display output in R)
```{r}
date_and_cite <- extract_from_esummary(multi_summs, c("pubdate", "pmcrefcount", "title"))
knitr::kable(head(t(date_and_cite)), row.names = FALSE)
```

## Fetching full records: `entrez_fetch()`

* As useful as the summary records are, sometimes they just don’t have the information that you need. 

* If you want a complete representation of a record you can use `entrez_fetch`, using the argument `rettype` to specify the format you’d like the record in.

# Fetch DNA sequences in fasta format

```{r}
gene_ids <- c(351, 11647)
linked_seq_ids <- entrez_link(dbfrom="gene", id=gene_ids, db="nuccore")
linked_transcripts <- linked_seq_ids$links$gene_nuccore_refseqrna
head(linked_transcripts)
```

* Now we can get our sequences with `entrez_fetch`, setting `rettype` to `“fasta”` (the list of formats available for each database is give in this [table](https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/)):

```{r}
all_recs <- entrez_fetch(db="nuccore", id=linked_transcripts, rettype = "fasta")
class(all_recs)
```
```{r}
nchar(all_recs)
```
* We now have a really huge character vector! 

* Rather than printing all those thousands of bases we can take a peak at the top of the file:

```{r}
cat(strwrap(substr(all_recs, 1, 500)), sep = "\n")
```
* If we wanted to use these sequences in some other application we could write them to file:

```{r}
write(all_recs, file = "my_transcripts.fasta")
```
* file will be saved in your current RStudio working directory

# End Lesson examples. 
* [exercise answers](exercise_answers.html)


### Topics not covered in this lesson:
* Fetch a parsed XML document
* Using NCBI's Web History features

## Where to learn more???
* Most of this tutorial was based on the [ROpenscience tutorial](https://ropensci.org/tutorials/rentrez_tutorial.html) on rentrez. 
* I also used some of Ted Hart's [Rentrez demo](https://github.com/ropensci/workshops-scio14-2014-02) as part of SIO14. 

